pkg/serverless/trace/inferredspan/propagation.go

```diff
@@ -0,0 +1,183 @@
+)
+
+const (
+	awsTraceHeader = "AWSTraceHeader"
+	datadogHeader  = "_datadog"
+
+	rootPrefix    = "Root="
+	parentPrefix  = "Parent="
+	rootPadding   = len(rootPrefix + "1-00000000-00000000")
+	parentPadding = len(parentPrefix)
+)
+
+var rootRegex = regexp.MustCompile("Root=1-[0-9a-fA-F]{8}-00000000[0-9a-fA-F]{16}")
+
+type rawTraceContext struct {
+	TraceID  string `json:"x-datadog-trace-id"`
+	ParentID string `json:"x-datadog-parent-id"`
+	base     int
+}
+
+type convertedTraceContext struct {
+	TraceID  *uint64
+	ParentID *uint64
+}
+
+type customMessageAttributeStruct struct {
+	Type  string `json:"Type"`
+	Value string `json:"Value"`
+}
+type bodyStruct struct {
+	MessageAttributes map[string]customMessageAttributeStruct `json:"MessageAttributes"`
+}
+
+func extractTraceContext(event events.SQSMessage) *convertedTraceContext {
+	var rawTrace *rawTraceContext
+
+	if awsAttribute, ok := event.Attributes[awsTraceHeader]; ok {
+		rawTrace = extractTraceContextfromAWSTraceHeader(awsAttribute)
+	}
+
+	if rawTrace == nil {
+		if ddMessageAttribute, ok := event.MessageAttributes[datadogHeader]; ok {
+			rawTrace = extractTraceContextFromPureSqsEvent(ddMessageAttribute)
+		} else {
+			rawTrace = extractTraceContextFromSNSSQSEvent(event)
+		}
+	}
+
+	return convertRawTraceContext(rawTrace)
+}
+
+func extractTraceContextFromSNSSQSEvent(firstRecord events.SQSMessage) *rawTraceContext {
+	var messageBody bodyStruct
+	err := json.Unmarshal([]byte(firstRecord.Body), &messageBody)
+	if err != nil {
+		log.Debug("Error unmarshaling the message body: ", err)
+		return nil
+	}
+
+	ddCustomPayloadValue, ok := messageBody.MessageAttributes["_datadog"]
+	if !ok {
+		log.Debug("No Datadog trace context found")
+		return nil
+	}
+
+	var traceData rawTraceContext
+	if ddCustomPayloadValue.Type == "Binary" {
+		decodedBinary, err := base64.StdEncoding.DecodeString(string(ddCustomPayloadValue.Value))
+		if err != nil {
+			log.Debug("Error decoding binary: ", err)
+			return nil
+		}
+		err = json.Unmarshal(decodedBinary, &traceData)
+		if err != nil {
+			log.Debug("Error unmarshaling the decoded binary: ", err)
+			return nil
+		}
+	} else {
+		log.Debug("Unsupported DataType in _datadog payload")
+		return nil
+	}
+
+	return &traceData
+}
+
+func extractTraceContextFromPureSqsEvent(ddPayloadValue events.SQSMessageAttribute) *rawTraceContext {
+	var traceData rawTraceContext
+	if ddPayloadValue.DataType == "String" {
+		err := json.Unmarshal([]byte(*ddPayloadValue.StringValue), &traceData)
+		if err != nil {
+			log.Debug("Error unmarshaling payload value: ", err)
+			return nil
+		}
+	} else {
+		log.Debug("Unsupported DataType in _datadog payload")
+		return nil
+	}
+
+	return &traceData
+}
+
+func extractTraceContextfromAWSTraceHeader(value string) *rawTraceContext {
+	if !rootRegex.MatchString(value) {
+		return nil
+	}
+
+	var startPart int
+	traceData := &rawTraceContext{base: 16}
+	length := len(value)
+	for startPart < length {
```

Comment:
Did we re-implement this algorithm from the java side? Is this canonical, etc?

Reply:
It's copied from where the java tracer extracts these headers: https://github.com/DataDog/dd-trace-java/blob/822ffc03d9317dfa4cbb6797038d600ba3659f83/dd-trace-core/src/main/java/datadog/trace/core/propagation/XRayHttpCodec.java#L191-L237

Reply:
Roger that

Reply:
I'm not sure what the best way ensure the insertion in the java tracer matches our extraction in the extension. I'm thinking adding an SQS tracing test app to self monitoring service is a good start though.